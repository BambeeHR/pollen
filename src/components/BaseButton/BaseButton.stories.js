import capitalize from 'lodash/capitalize';
import { action } from '@storybook/addon-actions';
import BaseButton from './BaseButton.vue';
import Button from '../../constants/Button';
import TypeOverline from '../TypeOverline/TypeOverline.vue';
import '../../assets/tailwind.css';

const { Sizes, Variants } = Button;

export default {
  title: 'BaseButton',
  component: BaseButton,
};

export const Gallery = () => ({
  components: { BaseButton, TypeOverline },
  data() {
    return {
      sizes: Sizes,
      variants: Variants,
    };
  },
  methods: {
    capitalize,
    handleClick(variant, size) {
      console.log(`clicked: ${variant} - ${size}`);
      action(`clicked: ${variant} - ${size}`); // TODO(jon.jandoc): This is not working.
    },
  },
  template: `
    <div>
      <TypeOverline tag="h1" variant="large">Regular buttons</TypeOverline>
      <div class="flex flex-wrap -mx-8 -mt-8">
        <div v-for="variant in variants" :key="variant" class="m-8 w-56">
        <TypeOverline tag="h2" class="mb-2">{{ variant }}</TypeOverline>
          <div v-for= "size in sizes" :key="'regular-' + variant + '-' + size" class="mb-2">
            <BaseButton :variant="variant" :size="size" @click="handleClick(variant, size)">{{ capitalize(size) }}</BaseButton>
          </div>
          <div v-for= "size in sizes" :key="'block-' + variant + '-' + size" class="mb-2">
            <BaseButton :variant="variant" :size="size" block @click="handleClick(variant, size)">{{ capitalize(size) }}</BaseButton>
          </div>
          <div>
            <BaseButton :variant="variant" block disabled @click="handleClick(variant, size)">Medium - Disabled</BaseButton>
          </div>
        </div>
      </div>
      <TypeOverline tag="h1" variant="large">Flat buttons</TypeOverline>
      <div class="flex flex-wrap -mx-8 -my-8">
        <div
          v-for="variant in variants" 
          :key="'flat' + variant"
          :class="variant === 'inverted' ? 'bg-gray-0 m-4 p-4 text-white w-48' : 'm-8 w-56'"
        >
          <TypeOverline tag="h2" class="mb-2">{{ variant }}</TypeOverline>
          <div v-for= "size in sizes" :key="'flat-regular-' + variant + '-' + size" class="mb-2">
            <BaseButton :variant="variant" :size="size" flat @click="handleClick(variant, size)">{{ capitalize(size) }}</BaseButton>
          </div>
          <div>
            <BaseButton :variant="variant" flat disabled @click="handleClick(variant, size)">Medium - Disabled</BaseButton>
          </div>
        </div>      
      </div>
    </div>
  `,
});

export const WithPreIcon = () => ({
  components: { BaseButton, TypeOverline },
  template: `
    <BaseButton preIcon="lock">Button</BaseButton>
  `,
});

export const WithPostIcon = () => ({
  components: { BaseButton, TypeOverline },
  template: `
    <BaseButton postIcon="lock">Button</BaseButton>
  `,
});
