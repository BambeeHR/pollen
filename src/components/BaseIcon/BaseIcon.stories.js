import BaseIcon, { Icons } from './BaseIcon.vue';
import '../../assets/tailwind.css';

export default {
  title: 'BaseIcon',
  component: BaseIcon,
};

export const Gallery = () => ({
  components: { BaseIcon },
  data() {
    return {
      icons: Object.values(Icons),
    };
  },
  template: `
    <div class="flex flex-wrap">
      <div v-for="icon in icons" :key="icon" :style="{ margin: '12px', textAlign: 'center' }">
        <BaseIcon :icon="icon" />
        <div :style="{ fontSize: '12px' }">{{ icon }}</div>
      </div>
    </div>
  `,
});
