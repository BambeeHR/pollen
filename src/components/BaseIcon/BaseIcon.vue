<template functional>
  <component
    :is="$options.getComponent(props.icon)"
    :class="[
      data.class,
      data.staticClass,
      { 'base-icon--custom': $options.isCustomIcon(props.icon) },
    ]"
    :icon="$options.getIcon(props.icon)"
    class="base-icon"
    v-bind="$options.getAttributes({ data, parent })"
    v-on="listeners"
  />
</template>

<script>
import camelCase from 'lodash/camelCase';
import snakeCase from 'lodash/snakeCase';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';

import {
  faFolder as folder,
  faSearch as search,
} from '@fortawesome/pro-regular-svg-icons';

import {
  faClipboardCheck as clipboardCheck,
  faPlus as plus,
  faTimes as close,
} from '@fortawesome/pro-light-svg-icons';

import {
  faCheck as check,
  faLock as lock,
} from '@fortawesome/free-solid-svg-icons';

import eyeClosed from '../../assets/icons/eye-closed.svg';
import eyeOpen from '../../assets/icons/eye-open.svg';

import getAttributes from '../../util/getAttributes';

const FAIcons = Object.freeze({
  check,
  clipboardCheck,
  close,
  folder,
  lock,
  search,
  plus,
});

Object.values(FAIcons).forEach((icon) => {
  library.add(icon);
});

const CustomIcons = Object.freeze({
  eyeClosed,
  eyeOpen,
});

export const Icons = Object.freeze(
  [...Object.keys(FAIcons), ...Object.keys(CustomIcons)].reduce(
    (obj, val) => ({
      ...obj,
      [snakeCase(val).toUpperCase()]: val,
    }),
    {}
  )
);

export const isValidIcon = (value) =>
  Object.values(Icons).includes(camelCase(value));

const isCustomIcon = (iconName) => Object.keys(CustomIcons).includes(iconName);

const getComponent = (iconName) =>
  isCustomIcon(iconName) ? CustomIcons[iconName] : FontAwesomeIcon;

/**
 * `import { BaseIcon } from '@bambeehr/pollen';`
 *
 * An SVG icon component pulling from Bambee's predefined icon set. Icons are
 * designed to be styled according to the contextual size and font color, e.g.
 * if you want a red, 40px icon, make sure it or its containing element is
 * styled with `color: red; font-size: 40px;`.
 */
export default {
  props: {
    /** The name of the icon to use. */
    icon: {
      type: String,
      required: true,
      validator: isValidIcon,
    },
  },
  getAttributes,
  getComponent,
  getIcon: (icon) => FAIcons[camelCase(icon)] || null,
  isCustomIcon,
};
</script>

<style>
.base-icon--custom {
  height: 1em;
  width: 1em;
}
</style>
