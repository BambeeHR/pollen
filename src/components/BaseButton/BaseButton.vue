<script>
import BaseIcon, { isValidIcon } from '../BaseIcon/BaseIcon.vue';
import Button from '../../constants/Button';
import getAttributes from '../../util/getAttributes';

const { Sizes, Variants } = Button;

export default {
  functional: true,
  components: { BaseIcon },
  props: {
    block: {
      type: Boolean,
      default: false,
    },
    flat: {
      type: Boolean,
      default: false,
    },
    preIcon: {
      type: String,
      default: null,
      validator: (value) => !value || isValidIcon(value),
    },
    postIcon: {
      type: String,
      default: null,
      validator: (value) => !value || isValidIcon(value),
    },
    size: {
      type: String,
      default: Sizes.MEDIUM,
      validator: (value) => Object.values(Sizes).includes(value),
    },
    tag: {
      type: String,
      default: 'button',
    },
    variant: {
      type: String,
      default: Variants.PRIMARY,
      validator: (value) => Object.values(Variants).includes(value),
    },
  },
  render(h, { data, props, parent, children }) {
    const preIcon = props.preIcon ? (
      <BaseIcon
        icon={props.preIcon}
        class="base-button--icon base-button--icon__pre"
      />
    ) : null;
    const postIcon = props.postIcon ? (
      <BaseIcon
        icon={props.postIcon}
        class="base-button--icon base-button--icon__post"
      />
    ) : null;
    const wrappedChildNodes = h('div', { class: 'base-button--inner' }, [
      preIcon,
      ...(children || []),
      postIcon,
    ]);

    return h(
      props.tag,
      {
        ...data,
        attrs: { ...getAttributes({ data, parent }) },
        class: [
          data.class,
          data.staticClass,
          'base-button',
          `base-button__${props.variant}`,
          `base-button__${props.size}`,
          {
            'base-button__flat': props.flat,
            'base-button__regular': !props.flat,
            'base-button__inline': !props.block || props.flat,
            'base-button__block': props.block && !props.flat,
          },
        ],
      },
      [wrappedChildNodes]
    );
  },
};
</script>

<style scoped>
.base-button {
  @apply border-transparent
    cursor-pointer
    items-center
    justify-center
    rounded
    whitespace-no-wrap
    /* For legacy */
    border-solid;
}

.base-button:focus {
  outline: 0;
}

.base-button[disabled] {
  @apply cursor-not-allowed opacity-50;
}

.base-button--inner {
  @apply flex
  font-semibold
    items-center
    justify-center
    tracking-wide
    w-full;
}

.base-button--icon__pre {
  @apply mr-2;
}

.base-button--icon__post {
  @apply ml-2;
}

/* Display variations */
.base-button__inline {
  @apply inline-flex;
}

.base-button__block {
  @apply flex w-full;
}

/* Size variations */
.base-button__small {
  @apply h-6 text-12;
}
.base-button__small .base-button--inner {
  @apply px-2;
}

.base-button__medium {
  @apply h-8 text-14;
}

.base-button__medium .base-button--inner {
  @apply px-4;
}

.base-button__large {
  @apply h-10 text-16;
}

.base-button__large .base-button--inner {
  @apply px-6;
}

.base-button__giant {
  @apply text-18;
  height: 3.5rem;
}

.base-button__giant .base-button--inner {
  @apply px-8;
}

/* Theme variations */
/** Regular buttons  */
.base-button__regular {
  @apply border-2 shadow-2;
}

/*** Regular primary buttons */
.base-button__regular.base-button__primary {
  @apply bg-primary text-white;
}

.base-button__regular.base-button__primary:focus {
  @apply border-primary-tint-3;
}

.base-button__regular.base-button__primary:active:not([disabled]) {
  @apply bg-primary-tint-1;
}

/*** Regular secondary buttons */
.base-button__regular.base-button__secondary {
  @apply bg-secondary text-white;
}

.base-button__regular.base-button__secondary:focus {
  @apply border-secondary-tint-3;
}

.base-button__regular.base-button__secondary:active:not([disabled]) {
  @apply bg-secondary-tint-1;
}

/*** Regular tertiary buttons */
.base-button__regular.base-button__tertiary {
  @apply bg-error text-white;
}

.base-button__regular.base-button__tertiary:focus {
  border-color: #e35b76; /* This is a one-off color. */
}

.base-button__regular.base-button__tertiary:active:not([disabled]) {
  background: #e32c50; /* This is a one-off color. */
}

/*** Regular primary inverted */
.base-button__regular.base-button__inverted {
  @apply bg-white border border-primary text-primary;
  padding: 1px;
}

.base-button__regular.base-button__inverted:focus {
  @apply border-2 p-0;
}

.base-button__regular.base-button__inverted:hover:not([disabled]) {
  @apply text-primary-tint-1;
}

.base-button__regular.base-button__inverted:active:not([disabled]) {
  @apply bg-primary-tint-5 text-primary;
}

/** Flat buttons */
.base-button__flat {
  @apply bg-transparent border;
}

/*** Flat primary buttons */
.base-button__flat.base-button__primary {
  @apply text-primary;
}

.base-button__flat.base-button__primary:focus {
  @apply border-primary-tint-3;
}

.base-button__flat.base-button__primary:hover:not([disabled]) {
  @apply text-primary-tint-1;
}

.base-button__flat.base-button__primary:active:not([disabled]) {
  @apply text-primary-shade-1;
}

/*** Flat secondary buttons */
.base-button__flat.base-button__secondary {
  @apply text-gray-0;
}

.base-button__flat.base-button__secondary:focus {
  @apply border-gray-5;
}

.base-button__flat.base-button__secondary:hover:not([disabled]) {
  @apply text-gray-2;
}

.base-button__flat.base-button__secondary:active:not([disabled]) {
  @apply text-black;
}

/*** Flat tertiary buttons */
.base-button__flat.base-button__tertiary {
  @apply text-error;
}

.base-button__flat.base-button__tertiary:focus {
  border-color: #f2dce1; /* This is a one-off color. */
}

.base-button__flat.base-button__tertiary:hover:not([disabled]) {
  color: #e32c50;
}

.base-button__flat.base-button__tertiary:active:not([disabled]) {
  color: #961d35; /* This is a one-off color. */
}

/*** Flat inverted buttons */
.base-button__flat.base-button__inverted {
  @apply text-white;
}

.base-button__flat.base-button__inverted:focus {
  border-color: rgba(255, 255, 255, 0.2);
}

.base-button__flat.base-button__inverted:hover:not([disabled]) {
  @apply text-gray-5;
}
</style>
