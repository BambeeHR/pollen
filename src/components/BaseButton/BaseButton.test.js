import { mount } from '@vue/test-utils';
import BaseButton from './BaseButton.vue';
import { Icons } from '../BaseIcon/BaseIcon.vue';

const testIcon = Icons.LOCK;

describe('BaseButton', () => {
  test('renders correctly', () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      propsData: {
        preIcon: testIcon,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders pre and post icons', () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      propsData: {
        preIcon: testIcon,
        postIcon: testIcon,
      },
    });
    expect(wrapper.findAll('.base-icon').length).toBe(2);
  });

  test('renders the component as a button by default', () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
    });
    expect(wrapper.element.tagName).toBe('BUTTON');
  });

  test('renders the component as a link when prop `tag` is `a`', () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      propsData: {
        tag: 'a',
      },
    });
    expect(wrapper.element.tagName).toBe('A');
  });

  test('transparently passes through classes and attributes', () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      attrs: {
        id: 'test-button',
        type: 'reset',
        disabled: true,
      },
    });
    expect(wrapper.element.id).toBe('test-button');
    expect(wrapper.element.type).toBe('reset');
    expect(wrapper.element.disabled).toBe(true);
  });

  test('emits a click event', () => {
    const onClick = jest.fn();
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      listeners: {
        click: onClick,
      },
    });
    wrapper.trigger('click');
    expect(onClick).toHaveBeenCalled();
  });

  test('does not emit a click event when disabled', () => {
    const onClick = jest.fn();
    const wrapper = mount(BaseButton, {
      slots: {
        default: 'Hello world',
      },
      listeners: {
        click: onClick,
      },
      attrs: {
        disabled: true,
      },
    });
    wrapper.trigger('click');
    expect(onClick).not.toHaveBeenCalled();
  });
});
