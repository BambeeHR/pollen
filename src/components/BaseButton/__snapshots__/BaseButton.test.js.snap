// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton renders correctly 1`] = `
<button
  class="base-button base-button--primary base-button--medium base-button--regular base-button--inline"
>
  <div
    class="base-button__inner"
  >
    <svg
      aria-hidden="true"
      class="base-icon svg-inline--fa fa-lock fa-w-14 base-button__icon base-button__icon--pre"
      data-icon="lock"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
        fill="currentColor"
      />
    </svg>
    <template>
      Hello world
    </template>
  </div>
</button>
`;
